designs/containers/vnf-deployment/vnf-deployment.md

```diff
@@ -84,23 +130,186 @@ public class ApplianceSoftwareVersion extends BaseEntity {
 }
 ```  
 
+#### Deployment Spec
+The `DeploymentSpec` domain entity will have some of its fields made `nullable`; the new `externalId` field to store the Kubernetes deployment UID; the `namespace` field for the deployment namespace; and set of the new entity `InspectionPort` :
+```java
+Entity
+@Table(name = "DEPLOYMENT_SPEC", ...)
+public class DeploymentSpec extends BaseEntity implements LastJobContainer {
+
+    // ...
+
+    @Column(name = "region", nullable = **true**)
+    private String region;
+
+    @Column(name = "project_name", nullable = **true**)
+    private String projectName;
+
+    @Column(name = "project_id", nullable = **true**)
+    private String projectId;
+
+    @Column(name = "management_network_name", nullable = **true**)
+    private String managementNetworkName;
+
+    @Column(name = "management_network_id", nullable = **true**)
+    private String managementNetworkId;
+
+    @Column(name = "inspection_network_name", nullable = **true**)
+    private String inspectionNetworkName;
+
+    @Column(name = "inspection_network_id", nullable = **true**)
+    private String inspectionNetworkId;
+	
+	@Column(name = "external_id", nullable = **true**)
+    private String externalId;                           // Stores the Kubernetes deployment id.
+	
+	@Column(name = "namespace", nullable = **true**)
+    private String namespace;                           // Stores the Kuberneted deployment namespace.
 	
+	@OneToMany(mappedBy = "deploymentSpec", fetch = FetchType.LAZY)
+    private Set<InspectionPort> inspectionPorts = new HashSet<InspectionPort>(); // The set of inspection ports associated with this DS.
+	// ...
+}
+```
+
+### Distributed Appliance Instances
+In Kubernetes, containers may be deleted and re-created at any time, without the direct control of OSC. This means it is possible for a container whose port is being used for traffic redirection to be deleted and replaced by Kubernetes. Because of that, OSC needs to have a way to update inspection ports registered with the SDN controller with a new virtual port. To enable this, OSC will persist two new columns in the entity `DistributedApplianceInsance`: `inspectionElementId` and `inspectionElementParentId`. These optional values will store the identifier of the registered inspection port and its parent respectively, both returned from the SDN controller method `SdnRedirectionApi.registerInspectionPort`. Also, the DAI will have a column to store the pod UID, `externalId`.  
+Additionally, most of the existing fields of the `DistributedApplianceInstance` entity will be reused for Kubernetes as indicated below: 
+
+
+```java
+@Entity
+@Table(name = "DISTRIBUTED_APPLIANCE_INSTANCE")
+public class DistributedApplianceInstance extends BaseEntity {
+	@Column(name = "name", unique = true, nullable = false)
+    private String name;                                     // Stores the name of the Kubernetes pod

```

Comment:
how is the uniqueness guaranteed when using deployments on k8s?

Reply:
pod names are unique on kubernetes and that is enforced.